### PR TITLE
transferFrom test

### PR DESCRIPTION
**Pull request by ilake**

### CODE REVIEW

**Comment by ilake** [comment]
@ngan This is good to review.

File: test/transferFrom.js [hunk 1]
```diff
@@ -3,13 +3,35 @@ const InkProtocol = artifacts.require("./mocks/InkProtocolMock.sol")
...
+    amount = 10
   })
 
   describe("#transferFrom()", () => {
-    it("fails when recipient is the protocol")
-    it("succeeds when recipient is another address")
+    it("fails when recipient is the protocol", async () => {
+      recipient = protocol.address
+
+      await $util.assertVMExceptionAsync(protocol.transferFrom(sender, recipient, amount))
```
Comment (ngan) [inline test/transferFrom.js]: Would this even succeed if the recipient was not the protocol?

File: test/transferFrom.js [hunk 2]
```diff
@@ -3,13 +3,35 @@ const InkProtocol = artifacts.require("./mocks/InkProtocolMock.sol")
...
+  let sender
+  let recipient
+  let agent
+  let amount
 
   beforeEach(async () => {
     protocol = await InkProtocol.new()
+    sender = accounts[1]
+    recipient = accounts[2]
+    agent = accounts[3]
```
Comment (ngan) [inline test/transferFrom.js:15]: `agent` does not apply to all tests, let's move this to just that one test that it's used in.

File: test/transferFrom.js [hunk 3]
```diff
@@ -3,13 +3,35 @@ const InkProtocol = artifacts.require("./mocks/InkProtocolMock.sol")
...
+    })
+
+    it("succeeds when recipient is another address", async () => {
+      await protocol.transfer(sender, amount)
+      let originalSenderBalance = await $util.getBalance(sender, protocol)
+
+      await protocol.approve(agent, amount, { from: sender })
+      await protocol.transferFrom(sender, recipient, amount, { from: agent })
+
+      assert.equal(await $util.getBalance(sender, protocol), originalSenderBalance - amount)
```
Comment (ngan) [inline test/transferFrom.js]: No need to lookup `originalSenderBalance` in line 28. You know it should be at the end: 0.

**Review by ngan — CHANGES_REQUESTED** [review]

File: test/transferFrom.js [hunk 4]
```diff
@@ -3,13 +3,35 @@ const InkProtocol = artifacts.require("./mocks/InkProtocolMock.sol")
 
 contract("InkProtocol", (accounts) => {
   let protocol
+  let sender
+  let recipient
+  let agent
+  let amount
```
Comment (ngan) [inline test/transferFrom.js]: Let's do:
```javascript
let protocol,
    sender,
    recipient,
    amount
```
to match the other tests

File: test/transferFrom.js [hunk 5]
```diff
@@ -3,13 +3,37 @@ const InkProtocol = artifacts.require("./mocks/InkProtocolMock.sol")
...
   describe("#transferFrom()", () => {
-    it("fails when recipient is the protocol")
-    it("succeeds when recipient is another address")
+    it("fails when recipient is the protocol", async () => {
+      recipient = protocol.address
+
+      await protocol.transfer(sender, amount)
+      await protocol.approve(agent, amount, { from: sender })
+
+      await $util.assertVMExceptionAsync(protocol.transferFrom(sender, recipient, amount, { from: agent }))
```
Comment (ngan) [inline test/transferFrom.js:26]: Let’s asset balances here too

**Review by ngan — CHANGES_REQUESTED** [review]

**Pull request merged**